[Dane Marc Yap Bagaoisan] iP

DaneMarc opened the pull request:
# Hola! soy ~~Dora~~ José
> "Backpack backpack" - Los Pollos Hermanos ([Source](https://www.smithsonianmag.com/arts-culture/the-history-of-the-frozen-banana-stand-80551198/))

José can get you anything from a taco to a gaucho but that's not all! He can:
- defend like a lion 🦁
- take care of your dear abuela
- eat

All you need to do is,
1. downloadido from [aqui](https://github.com/DaneMarc/ip/releases)
2. install
3. double-click
4. enjoy?

And it is **FREE**! That's right <span style="font-size:2rem">FREE</span>

Features:
- [x] we've 
- [ ] already
- [ ] done this

---
You Java programmer, you practice with `main` method like so:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

DaneMarc committed:
January 16 2022, 10:20pm - El nacimiento de José

DaneMarc committed:
Added list feature

DaneMarc committed:
Added Task class; list is now a list of tasks

DaneMarc committed:
Added ToDo, Deadline and Event

DaneMarc committed:
Updated testing

DaneMarc committed:
Added some exception handling

DaneMarc committed:
Added a delete function

DaneMarc committed:
Add saving and reading of task list from txt file

DaneMarc committed:
Add DateTime

DaneMarc committed:
Merge branch 'branch-Level-7'

DaneMarc committed:
Merge branch 'branch-Level-8'

DaneMarc committed:
Update project structure to be more OOPish

DaneMarc committed:
Update project structure

DaneMarc committed:
Add JUnit tests for the Task and TaskList classes

DaneMarc committed:
Update .gitignore

DaneMarc committed:
Add JavaDoc comments

DaneMarc committed:
Update code to comply with coding standard

DaneMarc committed:
Add find feature

DaneMarc committed:
Merge branch 'branch-A-JavaDoc'

DaneMarc committed:
Merge branch 'branch-A-CodingStandard'

DaneMarc committed:
Merge branch 'branch-Level-9'

yuqitanyq commented on src/main/java/jose/Duke.java:
```diff
@@ -0,0 +1,126 @@
+                    tempTask = tasks.getTask(parser.getIndex(input));
+                    tempTask.unmark();
+                    isModified = true;
+                    ui.showUnmarkMessage(tempTask);
+                    break;
+                case DELETE:
+                    tempTask = tasks.getTask(parser.getIndex(input));
+                    ui.showDeleteMessage(tempTask);
+                    tasks.removeTask(tempTask);
+                    isModified = true;
+                    ui.showRemainingTasks(tasks);
+                    break;
+                case FIND:
+                    taskInfo = input.split(" ", 2);
+                    ui.showList(tasks.findTasks(taskInfo[1]));
+                    break;
+                case TODO:
+                    taskInfo = input.split(" ", 2);
+                    tempTask = new ToDo(taskInfo[1]);
+                    tasks.addTask(tempTask);
+                    isModified = true;
+                    ui.showAddMessage(tempTask);
+                    ui.showRemainingTasks(tasks);
+                    break;
+                case DEADLINE:
+                    taskInfo = input.split(" ", 2)[1].split(" /by ");
+                    tempTask = new Deadline(taskInfo[0], taskInfo[1]);
+                    tasks.addTask(tempTask);
+                    isModified = true;
+                    ui.showAddMessage(tempTask);
+                    ui.showRemainingTasks(tasks);
+                    break;
+                case EVENT:
+                    taskInfo = input.split(" ", 2)[1].split(" /at ");
+                    tempTask = new Event(taskInfo[0], taskInfo[1]);
+                    tasks.addTask(tempTask);
+                    isModified = true;
+                    ui.showAddMessage(tempTask);
+                    ui.showRemainingTasks(tasks);
+                }
```
I like how you made the case clauses in the switch statements to have no indentation, which was highlighted in the Java coding standard. 👍

yuqitanyq commented on src/main/java/jose/Duke.java:
```diff
@@ -0,0 +1,126 @@
+    public void run() {
+        Scanner scanner = new Scanner(System.in);
+        Parser parser = new Parser();
+        boolean isModified = false;
+        boolean isRunning = true;
```
This naming of a boolean variable is good. 👍

yuqitanyq commented on src/main/java/jose/Storage.java:
```diff
@@ -0,0 +1,65 @@
+        file = new File(filePath);
+        if (!file.exists()) {
+            file.getParentFile().mkdirs();
+            file.createNewFile();
+        }
+    }
+
+    /**
+     * Reads the data file and compiles all the tasks saved in it for further processing.
+     *
+     * @return An ArrayList containing the tasks stored in the data file.
+     * @throws IOException If there are any errors accessing the data file.
+     */
+    public ArrayList<String> load() throws IOException {
+        BufferedReader br = new BufferedReader(new FileReader(file));
+        ArrayList<String> tasks = new ArrayList<>();
+        String s;
+
+        while ((s = br.readLine()) != null) {
+            tasks.add(s);
+        }
+
+        br.close();
+        return tasks;
+    }
+
+    /**
+     * Rewrites the data file with the given list of tasks.
+     *
+     * @param tasks  A list of tasks to be stored in the data file.
+     * @throws IOException If there are any errors when writing to the data file.
+     */
+    public void update(TaskList tasks) throws IOException {
+        BufferedWriter bw = new BufferedWriter(new FileWriter(file));
+        for (Task t : tasks.getTasks()) {
+            bw.write(t.formatData());
+            bw.newLine();
+        }
+        bw.close();
+    }
```
I like how clean this bit of code is! 👍

yuqitanyq commented on src/main/java/jose/Duke.java:
```diff
@@ -0,0 +1,126 @@
+package jose;
+
+import java.io.IOException;
+import java.util.Scanner;
+import jose.task.*;
```
I like how consistent your ordering of import statements are.

yuqitanyq commented on src/main/java/jose/TaskList.java:
```diff
@@ -0,0 +1,74 @@
+    public ArrayList<Task> findTasks(String query) {
+        ArrayList<Task> results = new ArrayList<>();
+
+        for (Task task : tasks) {
```
I like how there are spaces in between the loop conditions, but perhaps instead of using the syntactic sugar, the loop conditions could be specified. 😅

yuqitanyq commented on src/test/java/jose/TaskListTest.java:
```diff
@@ -0,0 +1,20 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class TaskListTest {
+    @Test
+    public void getSizeTest(){
+        assertEquals(0, new TaskList().getSize());
+    }
+
+    @Test
+    public void getTasksTest(){
+        assertEquals(new ArrayList<>(), new TaskList().getTasks());
+    }
```
I like how the tests are done and how simple it is to read! 👍

yuqitanyq reviewed (commented):
Overall, not many issues with the Java Coding standard! I found your code easy to read! Looks pretty good! I think this would make a great project and would think that the pull request can be accepted! LGTM.

LowJiaHao99 commented on src/main/java/jose/Parser.java:
```diff
@@ -0,0 +1,66 @@
+            } else if (command.equals("find")) {
+                return Command.FIND;
+            } else {
+                if (command.equals("todo") || command.equals("deadline") || command.equals("event")) {
```
Perhaps a switch statement can reduce deep nesting here?

LowJiaHao99 commented on src/main/java/jose/Storage.java:
```diff
@@ -0,0 +1,65 @@
+        ArrayList<String> tasks = new ArrayList<>();
+        String s;
+
+        while ((s = br.readLine()) != null) {
```
I like the way you section the code block, despite the fact that there are not many lines. The separation helps readers/developers understand that differences between those code blocks. 

I see many of such instances in the rest of the code. 👍

LowJiaHao99 commented on src/main/java/jose/Storage.java:
```diff
@@ -0,0 +1,65 @@
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.ArrayList;
```
Is there a way to sub-divide the import statements? How about I/O vs non-I/O related import statements?

DaneMarc replied:
Good idea. Didn't really think of that.

LowJiaHao99 commented on src/main/java/jose/task/Deadline.java:
```diff
@@ -0,0 +1,51 @@
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+
+public class Deadline extends Task {
```
Perhaps a header comment for JavaDocs regarding description of the class is missing?

I see that some of the classes also have such mistakes.

DaneMarc replied:
They are indeed missing. Thanks

LowJiaHao99 commented on src/main/java/jose/DukeException.java:
```diff
@@ -0,0 +1,15 @@
+/**
+ * A basic exception class that allows for more personalised error handling and messaging.
+ */
+public class DukeException extends Exception{
```
This might be a mistake, perhaps you have forgotten a white space between the class Exception and open parenthesis?

I notice  the same issue in other classes too.

DaneMarc replied:
Yep

LowJiaHao99 commented on src/main/java/jose/task/Deadline.java:
```diff
@@ -0,0 +1,51 @@
+    /**
+     * Constructor that sets all variables.
+     *
+     * @param description Task description
```
Do you think a punctuation is missing here?

I notice similar issues in the other classes.

LowJiaHao99 reviewed (commented):
Overall, it is pretty well done. Apart from a few minor mistakes, this,, in my opinion, is very impressive. The code is not deeply nested. Constants are defined for String to be printed. The methods are short and sweet. Import statements are mostly separated appropriately. JavaDoc comments are well maintained and the list goes on.

It is a pleasant experience to read this pull request. Good Job!! 👍  😄

DaneMarc committed:
Merge branch 'add-gradle-support'

DaneMarc committed:
Add checkstyle, update classes to comply with checkstyle and update build.gradle config

DaneMarc committed:
Add GUI

DaneMarc committed:
Merge branch 'branch-Level-10'

DaneMarc committed:
Add assert statements

User command information is assumed to be of the correct length after
parsing the input.

This might lead to Runtime Errors if user input is formatted wrongly.

Added assert statements after String splits to check if the given input
provides valid command information.

It is done this way to help us get familiarized with assert statements.

DaneMarc committed:
Refactor code

DaneMarc committed:
Merge pull request #1 from DaneMarc/branch-A-CodeQuality

Refactor code

DaneMarc committed:
Merge branch 'master' into branch-A-Assertions

DaneMarc committed:
Merge pull request #2 from DaneMarc/branch-A-Assertions

Add assert statements

DaneMarc committed:
Add task priority feature

Task objects don't contain a priority attribute so there was no way to
assign priority to a task.

Some tasks might be more important and warrant more attention from the
user or even the program.

Added the priority variable in the Task class and updated the other
classes and methods to accommodate this change.

It is done this way because the best and most structured way to
implement this change is to use the existing OOP structure.

DaneMarc committed:
Set theme jekyll-theme-modernist

DaneMarc committed:
Fix loads of bugs

DaneMarc committed:
Merge branch 'master' of https://github.com/DaneMarc/ip

DaneMarc committed:
Update README.md

DaneMarc committed:
Handle more exceptions and update README and demo screenshot
